Migrate to django-environ

tkdchen opened the pull request:
This patch uses django-environ and drop deprecated and old packages
django-cache-url, dj-email-url, dj-database-url and dj-cache-url.

Tests are updated according to the behavior of django-environ.

Signed-off-by: Chenxiong Qi <qcxhome@gmail.com>

tkdchen committed:
Migrate to django-environ

This patch uses django-environ and drop deprecated and old packages
django-cache-url, dj-email-url, dj-database-url and dj-cache-url.

Tests are updated according to the behavior of django-environ.

Signed-off-by: Chenxiong Qi <qcxhome@gmail.com>

blueyed commented on tests/test_values.py:
```diff
@@ -443,13 +439,7 @@ def test_cache_url_value(self):
-            self.assertEqual(
-                cm.exception.args[0],
-                "Cannot interpret cache URL value 'redis://user@host:port/1'")
+            self.assertEqual(cm.exception.args[0], 'wrong')
```
That looks "wrong"?!
Shouldn't it have more info / a full msg?

tkdchen replied:
django-environ simply raises `KeyError` when looking up a cache scheme by `wrong`, and here `assertRaises` catches the `KeyError`.

It might be good for django-envion to catch the KeyError and reraise a specific error with more descriptive message.

What do you think?

tkdchen replied:
FYI, I made a patch to django-environ, https://github.com/joke2k/django-environ/pull/235

jayvdb replied:
@tkdchen , that patch was merged; I guess this can be updated now ?

blueyed commented on tests/test_values.py:
```diff
@@ -411,26 +410,23 @@ def test_email_url_value(self):
             'default': {
                 'BACKEND': 'django_redis.cache.RedisCache',
-                'LOCATION': 'redis://host:6379/1',
+                'LOCATION': 'redis://user@host:6379/1',
```
Apparently it changes some defaults (more secure by default).

tkdchen replied:
I'm not sure if I get your point. Is this change acceptable?

blueyed replied:
It changes defaults apparently, so we should bump the major or at least minor version for it.

blueyed commented on tests/test_values.py:
```diff
@@ -411,26 +410,23 @@ def test_email_url_value(self):
-                'EMAIL_USE_SSL': False,
                 'EMAIL_USE_TLS': True})
-        with env(EMAIL_URL='console://'):
+        with env(EMAIL_URL='consolemail://'):
```
Is that change required?
Is it related to https://github.com/psgs/ConsoleMail?

tkdchen replied:
> Is it related to https://github.com/psgs/ConsoleMail?

No. I'm not aware of ConsoleMail. What I learned from django-environ is only `consolemail` is defined in email schemes, but no `console`. The `consolemail` maps to same backend `django.core.mail.backends.console.EmailBackend`.

blueyed replied:
That's another behavior change then apparently AFAIK.

All those might be fine, and it's good to modernize, but might break setups/configs.

tkdchen replied:
Hi @blueyed, I added another commit to make `console://` still work in apps using it, and deprecate `console://`.

blueyed commented on setup.py:
```diff
@@ -41,18 +41,11 @@ def find_version(*parts):
-            'dj-search-url',
             'six',
             'Sphinx>=1.4',
         ],
```
Could maybe still be in `extras_require`?

tkdchen replied:
Looks no need here. How useful to keep those in extra_require?

blueyed replied:
It avoids having to install / installing it by default when not needed.

I guess the typical usecase is using URLs for configs then, so that seems to be OK.

tkdchen committed:
Warn email schema console:// to be deprecated

Apps using schema console:// still works, which is converted to
consolemail:// internally in order to work with django-environ.
Meanwhile, a deprecation warning is output for console://.

Signed-off-by: Chenxiong Qi <qcxhome@gmail.com>

pauloxnet commented:
I think `django-environ` is a strong addiction. I would prefer to leave `django-configuration` as light as possible.

carltongibson commented:
Hi folks. 

Can you see also https://github.com/django/django/pull/13029 and the [related mailing list discussion](https://groups.google.com/d/topic/django-developers/CIPgeTetYpk/discussion)? 

**tl;dr**: Maybe we pull something into Django here. Perhaps it makes sense to have env vars read by the default project templates, but we should get right exactly what that should look like, and I don't want us to reinvent something that's already _solved™_ by the community. 

It would be good to get your thoughts as contributors here. 

Thanks 😉
C.

auvipy reviewed (changes requested):
I am very positive about this change moves o django-environ. can we just rebase and get this merged

Commit:
Merge branch 'master' into migrate-to-django-environ

brianhelba reviewed (changes requested):
@auvipy There are a some issues with your recent merge. See inline comments.

brianhelba reviewed (changes requested):
The docs for these Value types need to be updated too, as they contain references to the old upstream packages: https://github.com/jazzband/django-configurations/blob/master/docs/values.rst#url-based-values

Users looking for precise syntax info should be pointed to `django-environ` docs instead.

Commit:
Update setup.py

Co-authored-by: Brian Helba <brian.helba@kitware.com>

Commit:
Update setup.py

Co-authored-by: Brian Helba <brian.helba@kitware.com>

Commit:
Update setup.py

Co-authored-by: Brian Helba <brian.helba@kitware.com>

uhurusurfa commented:
I would agree with @pauloxnet on this one and would go a step further by creating an internal URL checker and converter to remove the dependencies on the 3 external packages. I had to fix a unit test that started failing due to a change in dj_database_url
I would propose closing this PR unless there remains strong opinions to move forward with the django-environ use.
